File: RightViewBinaryTree.java
// Time Complexity : O(n) where n is number of nodes in the input tree
// Space Complexity : O(h) where h is height of the binary tree
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

import java.util.*;
public class RightViewBinaryTree{
    public static void main(String[] args) {
        
    }
}
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if(root == null){
            return res;
        }
        dfs(root, res, 0);
        return res;
    }
    private void dfs(TreeNode root, List<Integer> res, int level){
        //base
        if(root == null){
            return;
        }
        
        // logic
        // since we are traversing right side of tree first if there is any value at the level, 
        // it means we already have the right most node in our resukt list
        if(level == res.size()){
            res.add(root.val);
        }
        dfs(root.right, res, level + 1);
        dfs(root.left, res, level + 1);
    }

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
}

File: rightViewBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

'''
Time Complexity -- > 

O(n)

Spacec Complexity -->

O(n) since we are using queue data structure
'''
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        if root==None:
            return res
        q = [root]
        res.append(root.val)
        while q:
            l = len(q)
            #print(l)
            ls = []
            for i in range(l):
                s = q.pop(0)
                #print(s.val)
                if s.left:
                    q.append(s.left)
                    ls.append(s.left.val)
                if s.right:
                    q.append(s.right)
                    ls.append(s.right.val)
            #print(q)
            if len(ls)!=0:
                res.append(ls[-1])
        return(res)
                
        